### PR TITLE
Controller:WaterCoil Maximum Actuated Flow should default to autosize and warn if zero

### LINKED ISSUE #7549: Controller:WaterCoil Maximum Actuated Flow should default to autosize and warn if zero
Issue overview
--------------
Simple user file with one chilled water coil showed zero chilled water flow all year. Eventually, tracked down the problem to this object.
```
Controller:WaterCoil,
    WATER FLOW CONTROLLER,   !- Name
    Temperature,             !- Control Variable
    Reverse,                 !- Action
    Flow,                    !- Actuator Variable
    cooling coil outlet node,!- Sensor Node Name
    CHILLED WATER INLET NODE,!- Actuator Node Name
    autosize,                !- Controller Convergence Tolerance {deltaC}
    ,                        !- Maximum Actuated Flow {m3/s}
    0.0000001;               !- Minimum Actuated Flow {m3/s}
```

Because Maximum Actuated Flow is blank and has no default in the IDD, it is zero, so the coil never gets any chilled water flow.

This field should default to autosize.
The controller should throw a warning if max flow is zero.

While fixing this - might want to confirm/address #4187 .

### Details
Some additional details for this issue (if relevant):
- Version of EnergyPlus v9.2.0
 - Helpdesk ticket number 15281

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added: EnergyPlusDevSupport\DefectFiles
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #7549
 - Make Max flow default to autosize, and warn if zero. Also made Min flow defaults to 0 (instead of  0.0000001, which is related to #4187)

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] ~If feature, test running new feature, try creative ways to break it~
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] ~Check any new function arguments for performance impacts~
 - [ ] ~Verify IDF naming conventions and styles, memos and notes and defaults~
 - [ ] ~If new idf included, locally check the err file and other outputs~

### CODE REVIEW

File: idd/Energy+.idd.in [hunk 1]
```diff
@@ -62646,12 +62646,15 @@ Controller:WaterCoil,
        \autosizable
   N2 , \field Maximum Actuated Flow
        \type real
+       \minimum 0
+       \autosizable
+       \default autosize
```
Comment (jmarrec) [inline idd/Energy+.idd.in]: New default for Maximum Actuated Flow

File: idd/Energy+.idd.in [hunk 2]
```diff
@@ -62646,12 +62646,15 @@ Controller:WaterCoil,
...
        \units m3/s
        \ip-units gal/min
-       \autosizable
   N3 ; \field Minimum Actuated Flow
        \type real
-       \default 0.0000001
+       \minimum 0
+       \default 0.0
```
Comment (jmarrec) [inline idd/Energy+.idd.in]: Changed default for Minimum Actuated Flow as well (cf #4187)

File: src/EnergyPlus/HVACControllers.cc [hunk 3]
```diff
@@ -1154,6 +1154,8 @@ namespace HVACControllers {
 
             // Check to make sure that the Minimum Flow rate is less than the max.
             if (ControllerProps(ControlNum).MaxVolFlowActuated == 0.0) {
+                ShowWarningError("Controller:WaterCoil, Maximum Actuated Flow is zero; " +
+                                 ControllerProps(ControlNum).ControllerName);
```
Comment (jmarrec) [inline src/EnergyPlus/HVACControllers.cc]: New warning

File: tst/EnergyPlus/unit/HVACControllers.unit.cc [hunk 4]
```diff
@@ -776,4 +781,245 @@ TEST_F(EnergyPlusFixture, HVACControllers_CheckTempAndHumRatCtrl)
...
+        " Controller:WaterCoil,",
+        "   CW Coil Controller, !- Name",
+        "   HumidityRatio,      !- Control Variable",
+        "   Reverse,            !- Action",
+        "   FLOW,               !- Actuator Variable",
+        "   Air Outlet Node,    !- Sensor Node Name",
+        "   Water Inlet Node,   !- Actuator Node Name",
+        "   ,                   !- Controller Convergence Tolerance { deltaC }",
+        "   ,                   !- Maximum Actuated Flow { m3 / s }",
+        "   ;                   !- Minimum Actuated Flow { m3 / s }",
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/HVACControllers.unit.cc:812]: blanks

File: tst/EnergyPlus/unit/HVACControllers.unit.cc [hunk 5]
```diff
@@ -776,4 +781,245 @@ TEST_F(EnergyPlusFixture, HVACControllers_CheckTempAndHumRatCtrl)
...
+    ASSERT_TRUE(process_idf(idf_objects));
+
+    GetSetPointManagerInputs();
+
+    GetControllerInput();
+
+    ASSERT_EQ(ControllerProps.size(), 1u);
+    EXPECT_EQ(ControllerProps(1).MaxVolFlowActuated, DataSizing::AutoSize);
+    EXPECT_EQ(ControllerProps(1).Offset, DataSizing::AutoSize);
+    EXPECT_EQ(ControllerProps(1).MinVolFlowActuated, 0.0);
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/HVACControllers.unit.cc:845]: Check that they end up with the right default.

File: tst/EnergyPlus/unit/HVACControllers.unit.cc [hunk 6]
```diff
@@ -776,4 +781,245 @@ TEST_F(EnergyPlusFixture, HVACControllers_CheckTempAndHumRatCtrl)
...
+}
+
+TEST_F(EnergyPlusFixture, HVACControllers_MaxFlowZero)
+{
+    std::string const idf_objects = delimited_string({
+        " Coil:Cooling:Water,",
+        "   Chilled Water Coil, !- Name",
+        "   AvailSched,         !- Availability Schedule Name",
+        "   0.00,               !- Design Water Flow Rate { m3 / s }",
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/HVACControllers.unit.cc:854]: CHW Coil is zero

File: tst/EnergyPlus/unit/HVACControllers.unit.cc [hunk 7]
```diff
@@ -776,4 +781,245 @@ TEST_F(EnergyPlusFixture, HVACControllers_CheckTempAndHumRatCtrl)
...
+    Real64 expectedOffset = (0.001 / (2100.0 * DataHVACGlobals::SmallWaterVolFlow)) * (DataConvergParams::HVACEnergyToler / 10.0);
+    // do not let the controller tolerance exceed 1/10 of the loop temperature tolerance.
+    expectedOffset = min(0.1 * DataConvergParams::HVACTemperatureToler, expectedOffset);
+    EXPECT_EQ(expectedOffset, 0.1 * DataConvergParams::HVACTemperatureToler);
+    EXPECT_EQ(ControllerProps(1).Offset, expectedOffset);
+
+    std::string const error_string = delimited_string({
+      "   ** Warning ** Controller:WaterCoil, Maximum Actuated Flow is zero; CW COIL CONTROLLER",
+    });
+    EXPECT_TRUE(compare_err_stream(error_string, true));
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/HVACControllers.unit.cc:1021]: Warning is issued.

File: tst/EnergyPlus/unit/HVACControllers.unit.cc [hunk 8]
```diff
@@ -776,4 +781,245 @@ TEST_F(EnergyPlusFixture, HVACControllers_CheckTempAndHumRatCtrl)
...
+    // and issue a warning
+    SimAirServingZones::SimAirLoops(true, SimZoneEquipment);
+
+    // after controllers are simulated, AirLoopControllerIndex = index to this controller on this air loop (e.g., n of num contollers on air loop)
+    ASSERT_EQ(1, DataAirSystems::PrimaryAirSystem(1).NumControllers);
+    ASSERT_EQ(1, DataAirSystems::PrimaryAirSystem(1).ControllerIndex(1));
+    ASSERT_EQ(1, ControllerProps(1).AirLoopControllerIndex);
+
+    // This should have been autosized to zero
+    EXPECT_EQ(ControllerProps(1).MaxVolFlowActuated, 0.0);
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/HVACControllers.unit.cc:1008]: Autosized to the CHW Coil design flow rate, that is zero.

**Comment by nrel-bot-2** [comment]
@jmarrec @lgentile it has been 28 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@jmarrec @lgentile it has been 29 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
@jmarrec I pulled upstream/develop into here to get a full set of good results.

**Comment by jmarrec** [comment]
Err diffs are all expected (I checked all 7):
```
+   ** Warning ** Controller:WaterCoil, Maximum Actuated Flow is zero; CENTRAL COOLING COIL CONTROLLER 1
```

eg: http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2020-02/55f5f74c4f7b5b99e0110e5c7a158fd85bb20cc2-91d042bc125dce9e0862d78e8f3b3c923466b84b/HospitalBaseline/x86_64-MacOS-10.13-clang/eplusout.err.diff.html

**Comment by jmarrec** [comment]
@Myoldmopar I saw that the custom_check was giving (weird, "Unknownfile:XXXX") warnings on clang_format, so I tried to apply clang-format to the changed files, but it doesn't seem like it made much of a difference...

**Comment by Myoldmopar** [comment]
Ignore it, I already have it fixed here: https://github.com/NREL/EnergyPlus/pull/7778

File: idd/Energy+.idd.in [hunk 9]
```diff
@@ -62666,12 +62666,15 @@ Controller:WaterCoil,
...
        \type real
+       \minimum 0
+       \autosizable
+       \default autosize
        \units m3/s
        \ip-units gal/min
-       \autosizable
   N3 ; \field Minimum Actuated Flow
        \type real
-       \default 0.0000001
```
Comment (mjwitte) [inline idd/Energy+.idd.in]: @Myoldmopar This default has been in the idd as far back as version 1.2.0, so my guess is that it was the idea of the original controller author.

**Comment by Myoldmopar** [comment]
With the nonzero default in there since the very beginning of the model, it's hard to believe we can just change that value without any ramifications.  With our IO freeze basically now, and major other PRs to get in, I don't see how this can be investigated fully and reviewed.  I'm going to push this to the next milestone and we'll have some time to get it figured out fully.

**Comment by nrel-bot-2** [comment]
@jmarrec it has been 22 days since this pull request was last updated.

**Comment by nrel-bot-2b** [comment]
@jmarrec it has been 7 days since this pull request was last updated.

**Comment by jmarrec** [comment]
Rebased onto develop and force-pushed

**Review by mjwitte — COMMENTED** [review]
Tested the 7549 defect file before and after this fix. The controller max flow (which is blank in the defect file) now autosizes to match the ChW coil max flow rate.
Modifying the defect file with a zero in the max flow rate field generates the new warning.

Several test suite files are now throwing the zero flow warning which is appropriate, because various coils are sizing to zero flow (e.g. main cooling coil where there is an OA cooling coil - may not be the best but that's how it's been, nothing changed here) and some hot water heating coils are also showing the new warning.

The only concern here is that the warning format is not consistent with existing warnings.
`   ** Warning ** Controller:WaterCoil, Maximum Actuated Flow is zero; WATER FLOW CONTROLLER`
but it does match the format of the min>max error message just a couple of lines later.

@Myoldmopar @rraustad Do we care about the warning format? We don't have a standard, but most places would be more like `   ** Warning ** Controller:WaterCoil = WATER FLOW CONTROLLER, Maximum Actuated Flow is zero.`

**Review by mjwitte — APPROVED** [review]
I reformatted the new warning and the error below it. Will wait for CI then merge.

**Pull request merged**